src/librustc/ty/print/pretty.rs

```diff
@@ -919,7 +919,10 @@ pub trait PrettyPrinter<'tcx>:
             },
             (ConstValue::Scalar(Scalar::Raw { data, .. }), ty::Char) =>
                 p!(write("{:?}", ::std::char::from_u32(data as u32).unwrap())),
-            (ConstValue::Scalar(_), ty::RawPtr(_)) => p!(write("{{pointer}}")),
+            (ConstValue::Scalar(Scalar::Raw { data, size }), ty::RawPtr(_)) => {
```

Comment:
Why would doing this depend on the type? At least since seems just as reasonable for references as it does for raw pointers.

Reply:
Are you saying we should print the address for const refs as well? Are they supposed to be comparable to each other? I currently get "mismatched types" when I try to equate two references to the same variable:
```rust
#![feature(const_generics)]

const A: u32 = 3;

struct Const<const P: &'static u32>;

fn main() {
    let _: Const<{&A}> = Const::<{&A}>;
}
```

code above gives me error below
```
error[E0308]: mismatched types
  --> /home/iwikal/code/rust/src/test/ui/const-generics/ref-const-param.rs:12:26
   |
LL |     let _: Const<{&A}> = Const::<{&A}>;
   |                          ^^^^^^^^^^^^^ expected `Scalar(AllocId(2).0x0) : &u32`, found `Scalar(AllocId(4).0x0) : &u32`
   |
   = note: expected type `Const<Scalar(AllocId(2).0x0) : &u32>`
              found type `Const<Scalar(AllocId(4).0x0) : &u32>`
```

(also note the extra space before the colon)

If this is the intended behaviour, then it might be confusing to print the address - it suggests that the address identifies them, when really they are all unique.

Reply:
Ah, what you are seeing here has nothing to do with references vs. raw pointers. This is the difference between pointers that point to anything vs pointer creates from an integer: the former are represented as pairs of an abstract allocation ID and an offset, while the latter are just raw integers. You can get `Scalar(AllocId(2).0x0)` in raw pointers by casting a reference to a raw ptr, and you can get `Scalar::Raw` in a reference by casting a raw ptr to a reference.

Reply:
Ah, interesting, thanks! I'll match refs in the type pattern as well, and add a test.

Reply:
I'm having trouble creating such a reference. From what I can gather, a reference must never ever be dangling, because that will always constitute undefined behavior. So how exactly would I create a reference from a literal address?

Reply:
I think you need to "unleash miri" to do it. See https://github.com/rust-lang/rust/tree/master/src/test/ui/consts/miri_unleashed for some existing tests that do that.

Reply:
```
error[E0080]: it is undefined behavior to use this value
  --> /home/iwikal/code/rust/src/test/ui/const-generics/ref-const-param.rs:12:33
   |
LL |     let _: Const<{A}> = Const::<{B}>; //~ mismatched types
   |                                 ^^^ type validation failed: encountered dangling reference (created from integer)
   |
   = note: The rules on what exactly is undefined behavior aren't clear, so this check might be overzealous. Please open an issue on the rustc repository if you believe it should not be considered undefined behavior.
```
This seems like it expressly forbids what I'm trying to do.

Reply:
Try using a reference to a ZST. Those are actually allowed to be dangling.

Also, thinking about this again, you should be able to transmute an integer to a reference through a `union` without "miri unleashed".